### PR TITLE
Spam comment moderation 

### LINKED ISSUE #2304: PLANNING ISSUE: Spam comments moderation
A user when banned by moderators, the alert text said that banning the author would remove all their content from the site -- but it did not. This is a bug.So, for handling this a comment moderation system needs to be built as discussed in #1785 .

* [x] make a "status" column for comments, default `1`, spam = `0`
* [x] make a "spam" method like this, that only moderators or admins can use (`user.role == admin`): https://github.com/publiclab/plots2/blob/master/app/models/node.rb#L158
* [x] a "publish" method which sets `status` back to 1
* [x] unit test the above 2 methods
* [x] display comments (on dashboard, /research, in answers, wiki pages and notes) only if they have `status == 1`
* [ ] display spammed comments to admins and moderators (`user.role == admin || user.role...`) but with text hidden, with a notice "A comment by @______ was put in moderation. _Read the full comment_" which displays the text, and a "republish" button

### PR DESCRIPTION
**Pull request by Gauravano**
Fixes #2304 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] PR body includes `fixes #0000`-style reference to original issue #
* [x] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. 

Thanks!

### CODE REVIEW

File: Dangerfile [hunk 1]
```diff
@@ -59,5 +59,5 @@ begin
 
 rescue => ex
   fail "There was an error with Danger bot's Junit parsing: #{ex.message}"
-  puts ex.inspect # view the entire error output in the log
+  message ex.inspect # view the entire error output in the log
```
Comment (Gauravano) [inline Dangerfile]: Just to handle warning!

File: app/controllers/admin_controller.rb [hunk 2]
```diff
@@ -131,7 +131,9 @@ def mark_comment_spam
     if current_user && (current_user.role == 'moderator' || current_user.role == 'admin')
       if @comment.status == 1
         @comment.spam
-        flash[:notice] = "Comment has been marked as spam."
+        user = @comment.author
+        user.ban
```
Comment (Gauravano) [inline app/controllers/admin_controller.rb]: What do you think should we moderate user or ban it ?

File: app/views/notes/_comment.html.erb [hunk 3]
```diff
@@ -42,9 +42,16 @@
...
           <a href="#" onclick="return false;" class="btn btn-default btn-sm"><i class="fa fa-envelope"></i></a>
         <% end %>
-        <a rel="tooltip" title="Flag as spam" class="btn btn-sm btn-default btn-flag-spam-<%= comment.id %>" href="mailto:moderators@publiclab.org?subject=Reporting+spam+on+Public+Lab&body=Hi,+I+found+this+comment+that+looks+like+spam+or+needs+to+be+moderated:+https://publiclab.org/<%= comment.parent.path %>#c<%= comment.cid %>+by+https://publiclab.org/profile/<% if comment.author %><%= comment.author.name %><% end %>+Thanks!">
-          <i class="fa fa-flag"></i>
-        </a>
+      <% if current_user &. can_moderate? %>
+          <a rel="tooltip" title="Mark as spam" class="btn btn-sm btn-default btn-flag-spam-<%= comment.id %>" data-confirm="Are you sure? The user will no longer be able to log in or publish, and their content will be hidden except comments." href="/admin/mark_comment_spam/<%= comment.id %>">
+            <i class="fa fa-ban"></i>
+          </a>
+    <% else %>
```
Comment (Gauravano) [inline app/views/notes/_comment.html.erb]: I have replaced "Flag as spam" with "mark as spam" for moderators and admins.

**Comment by Gauravano** [comment]
@jywarren please review this one. I have left some comments above please see if I have started right as after that, the same links would be added to spam page and comment tab of the profile, so just confirming. Thanks!

File: app/controllers/admin_controller.rb (hunk 2 above)
Comment (jywarren) [reply app/controllers/admin_controller.rb]: Ban is right - moderation is for a code of conduct violation. But we should better explain this in documentation somewhere! For example the buttons shown on an admin's view of others' profiles is confusing - both are shown as options!

File: app/views/notes/_comment.html.erb (hunk 3 above)
Comment (jywarren) [reply app/views/notes/_comment.html.erb]: 👍

**Comment by jywarren** [comment]
Looks like this drifted out of sync -- sorry, i just wanted to check with you that it's merge-able -- once you resolve the conflict, can you mark it 'ready'? Thank you @Gauravano !!! 👍

**Comment by Gauravano** [comment]
@jywarren I cleared conflicts but we can't merge this until its confirmed that all comments are migrated to status = 1.

**Comment by Gauravano** [comment]
@jywarren we can merge this one now :grinning:

**Comment by jywarren** [comment]
Hooray!!! 🎉

**Pull request merged**